content/blog/2023-12-11-circ-bitwidth-adjustment

```diff
@@ -0,0 +1,115 @@
+3. We have also made changes to the constraint generation system to take advantage of the two new operations. In the original implementation, the constraint generation system will always adjust the bitwidth of the result of an operation. We have made changes to the system to not adjust `add_unadjust` and `mul_unadjust` operations. We also proved that such change will not impact the correctness of the system.
+
+## What were the hardest parts to get right
+The hardest part is to make sure the compiler works properly using our newly introduced operation. We encounter many bugs while implementing the optimization. For instance, before since the author always adjust the bitwidth after all types of operations, the bitwidth is fixed. However, now since an operation might be using unadjusted bitwidth the logic of calculating the bitwidth of the current element is more complicated and requires more steps to ensure an operation does not overflow.
+
+## Empirical evaluation
+We uses two types of tests. 
+1. We use the existing benchmarks in the circ repo to test whether the output of the optimized compiler is correct. Our implementation passed all the test cases in CirC.
+2. To measure how well the proposed design improves the system, we use the number of r1cs constraints generated as the metrics. We runs our compiler and the original one over a set of zokrates programs in [CirC](https://github.com/circify/circ) and [Zombie](https://github.com/PepperSieve/Zombie/tree/master) and compares the number of constraints generated. Here is a shorted list of tests we run. The first line are the numbers of constraints with our flexible adjustment implementation and the second line is the result using original compiler.
+```
+../Zombie/circ/zkmb/shaRound.zok
+with optimization:    27137
+without optimization: 29057
+../Zombie/circ/zkmb/DotChaChaAmortized_Isolated.zok
+with optimization:    23366
+without optimization: 41835
+../Zombie/circ/zkmb/DotChaChaAmortized.zok
+with optimization:    111529
+without optimization: 130232
+../Zombie/circ/zkmb/tls_key_schedules/HKDF.zok
+with optimization:    101304
+without optimization: 109187
+../Zombie/circ/zkmb/SHA.zok
+with optimization:    26581
+without optimization: 28505
+../Zombie/circ/zkmb/DotChaChaAmortizedUnpack.zok
+with optimization:    31422
+without optimization: 50105
+```
+
+## How to run the code
+To run the r1cs compiler, first clone the repo, and checkout our branch
+```
+git clone https://github.com/collinzrj/circ
+git checkout bitwidth_adjustment
+```
+then `cd` into it,
+then run this command to build circ
+`cargo build --release --features r1cs,smt,zok,bellman --example circ`
+For proving and verifying you also need to build zk
+`cargo build --release --features r1cs,smt,zok,bellman --example zk`
+You can find them built in `./target/release/examples/zk` and `./target/release/examples/circ`
```

Comment:
I think these operational instructions belong in the README for your repo. They're not that relevant to the notional reader for this blog post, who is someone who just wants to understand what your scientific contribution was.